File: python/d06.py
#!/usr/bin/env python3

import util

def get_edges():
    inputs = util.get_puzzle_input().split('\n')

    edges = []
    for i in inputs:
        if i:
            i = i.split(')')
            edges.append( (i[0],i[1]) )
    return set(edges)

def part_a():
    edges = get_edges()

    distance = {'COM': 0}
    while edges:
        found = set()
        for a,b in edges:
            if a in distance:
                distance[b] = distance[a]+1
                found.add( (a,b) )
        edges = edges - found

    print('Total direct+indirect orbits:', sum(distance.values()))


def part_b():
    edges = get_edges()

    parent = {}
    for a,b in edges:
        parent[b] = a

    def path_to_com(start):
        path = [start]
        while path[-1]!='COM':
            path.append(parent[path[-1]])
        return path

    you_path = set(path_to_com('YOU'))
    san_path = set(path_to_com('SAN'))

    transfer_path = (you_path | san_path) - (you_path & san_path)
    # If transfer path is just 'YOU' and 'SAN', already there.
    # Otherwise, each additional step is one transfer
    print('Number of transfers:',len(transfer_path)-2)

def main():
    part_a()
    part_b()

if __name__ == '__main__':
    main()

File: cpp/d06.cc
#include <iostream>
#include <map>
#include <string>
#include <vector>

#include "ReadInputs.hh"

using Edge = std::pair<std::string, std::string>;


void part_a(const std::vector<Edge>& edges) {
  std::map<std::string, int> distance;
  distance["COM"] = 0;

  while(distance.size() != edges.size()+1) {
    for(auto edge : edges) {
      if(distance.count(edge.first) && !distance.count(edge.second)) {
        distance[edge.second] = distance[edge.first] + 1;
      }
    }
  }

  int sum = 0;
  for(auto pair : distance) {
    sum += pair.second;
  }
  std::cout << "Total direct+indirect orbits: " << sum << std::endl;
}


void part_b(const std::vector<Edge>& edges) {
  std::map<std::string, std::string> parent;
  for(const auto& edge : edges) {
    parent[edge.second] = edge.first;
  }

  auto path_to_com = [&](std::string start) {
    std::vector<std::string> path{start};
    while(start != "COM") {
      start = parent[start];
      path.push_back(start);
    }
    return path;
  };

  auto you_path = path_to_com("YOU");
  auto san_path = path_to_com("SAN");

  while(you_path.back() == san_path.back()) {
    you_path.pop_back();
    san_path.pop_back();
  }
  int num_transfers = you_path.size() + san_path.size() - 2;
  std::cout << "Number of transfers: " << num_transfers << std::endl;
}


int main(int argc, char** argv){
  if(argc < 2) {
    std::cerr << "Missing argument: puzzle input" << std::endl;
    return 1;
  }

  std::string puzzle_input_file(argv[1]);
  auto lines = read_lines(puzzle_input_file);



  std::vector<Edge> edges;
  for(auto line : lines) {
    std::string parent = line.substr(0,3);
    std::string child = line.substr(4,7);
    edges.push_back({parent, child});
  }
  part_a(edges);
  part_b(edges);
}
